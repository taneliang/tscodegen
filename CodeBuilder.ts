import prettier from 'prettier';

export class CodeBuilder {
  private gennedCode = '';

  readonly #existingManualSections: { [sectionKey: string]: string };

  constructor(manualSections: { [sectionKey: string]: string }) {
    this.#existingManualSections = manualSections;
  }

  insertCode(code: string): this {
    this.gennedCode += code;
    return this;
  }

  addManualSection(
    sectionKey: string,
    sectionBuilder: (manualSectionBuilder: CodeBuilder) => CodeBuilder,
  ) {
    // Ensure section key contains no whitespaces
    if (/\s/.test(sectionKey)) {
      throw new Error(
        `Manual section keys should not contain whitespaces. Received "${sectionKey}", which does.`,
      );
    }

    this.gennedCode += `\n/* BEGIN MANUAL SECTION ${sectionKey} */\n`;

    if (this.#existingManualSections[sectionKey]) {
      this.gennedCode += this.#existingManualSections[sectionKey];
    } else {
      this.gennedCode += sectionBuilder(new CodeBuilder(this.#existingManualSections)).toString();
    }

    this.gennedCode += `\n/* END MANUAL SECTION */\n`;
    return this;
  }

  format(): this {
    // TODO: Call Prettier
    this.gennedCode = prettier.format(this.gennedCode, { parser: 'typescript' });
    return this;
  }

  toString() {
    return this.gennedCode;
  }
}
