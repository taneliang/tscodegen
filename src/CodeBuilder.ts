import prettier from "prettier";
import { createManualSection } from "./sections/manual";
import type { ManualSectionMap } from "./types/ManualSectionMap";
import { createDocblock } from "./sections/docblock";

export class CodeBuilder {
  #gennedCode = "";
  #hasManualSections = false;

  readonly #existingManualSections: ManualSectionMap;

  constructor(manualSections: ManualSectionMap) {
    this.#existingManualSections = manualSections;
  }

  /**
   * Simply appends `code`.
   */
  add(code: string): this {
    this.#gennedCode += code;
    return this;
  }

  /**
   * Appends `code` and a newline. Call without arguments to insert a newline.
   */
  addLine(code: string = ""): this {
    this.#gennedCode += code + "\n";
    return this;
  }

  /**
   * Creates a docblock from `docblockContent` and appends it, with a trailing
   * newline.
   *
   * @param docblockContent Plain docblock content (i.e. without "*"s at the start of each line)
   */
  addDocblock(docblockContent: string): this {
    return this.addLine(createDocblock(docblockContent));
  }

  /**
   * Add a block of code, i.e. code with braces around them.
   *
   * @param codeBeforeBlock Code before the block's `{`, e.g. `if (a === b)`
   * @param blockBuilder A function that uses `blockBuilder` to build the code
   * in the block.
   */
  addBlock(
    codeBeforeBlock: string,
    blockBuilder: (blockBuilder: CodeBuilder) => CodeBuilder
  ): this {
    const builtBlockBuilder = blockBuilder(
      new CodeBuilder(this.#existingManualSections)
    );
    this.#hasManualSections =
      this.#hasManualSections || builtBlockBuilder.hasManualSections();
    return this.add(codeBeforeBlock)
      .addLine(" {")
      .addLine(builtBlockBuilder.toString())
      .addLine("}");
  }

  /**
   * Add a section of code that can be manually edited. The section will be
   * surrounded by "BEGIN MANUAL SECTION" and "END MANUAL SECTION" designators.
   *
   * @param sectionKey An identifier for this manual section.
   * @param blockBuilder A function that uses `blockBuilder` to build the code
   * in the block, if no existing manual section with `sectionKey` is stored.
   */
  addManualSection(
    sectionKey: string,
    sectionBuilder: (manualSectionBuilder: CodeBuilder) => CodeBuilder
  ): this {
    let sectionContent: string;
    if (this.#existingManualSections[sectionKey]) {
      sectionContent = this.#existingManualSections[sectionKey];
    } else {
      sectionContent = sectionBuilder(
        new CodeBuilder(this.#existingManualSections)
      ).toString();
    }
    this.#hasManualSections = true;
    return this.addLine(createManualSection(sectionKey, sectionContent));
  }

  /**
   * Formats the stored code with Prettier.
   */
  format(): this {
    const prettierOptions = prettier.resolveConfig.sync(process.cwd());
    this.#gennedCode = prettier.format(this.#gennedCode, {
      ...(prettierOptions ?? {}),
      parser: "typescript",
    });
    return this;
  }

  /**
   * Returns the stored code.
   */
  toString(): string {
    return this.#gennedCode;
  }

  /**
   * Whether this built code contains at least one manual section.
   */
  hasManualSections(): boolean {
    return this.#hasManualSections;
  }
}
