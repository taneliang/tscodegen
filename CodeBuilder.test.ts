import { CodeBuilder } from './CodeBuilder';

describe(CodeBuilder, () => {
  describe(CodeBuilder.prototype.insertCode, () => {
    test('should insert code', () => {
      expect(new CodeBuilder({}).insertCode('const hello = "world";').toString()).toBe(
        'const hello = "world";',
      );
    });
  });

  describe(CodeBuilder.prototype.addManualSection, () => {
    test('should generate manual section using a new builder', () => {
      const mockSectionBuilder = jest.fn().mockImplementation(() => ({
        toString: () => 'console.log("New manual section");',
      }));
      const builder = new CodeBuilder({
        someOtherManualSectionKey: 'ANOTHER EXISTING MANUAL SECTION; SHOULD NOT APPEAR',
      });
      builder.addManualSection('mansec', mockSectionBuilder);

      // Expect manual section to be built correctly, i.e. with content of
      // section builder, with the provided section name and valid demarcations.
      expect(builder.toString()).toMatchSnapshot();

      // Expect a new builder to be used for manual section.
      expect(mockSectionBuilder).toBeCalledTimes(1);
      expect(mockSectionBuilder).not.toBeCalledWith(builder);
    });

    test('should retain existing manual section content if present', () => {
      const mockSectionBuilder = jest.fn().mockImplementation(() => ({
        toString: () => 'NEW MANUAL SECTION; SHOULD NOT APPEAR',
      }));
      const builder = new CodeBuilder({
        mansec: 'console.log("Existing manual section");',
      });
      builder.addManualSection('mansec', mockSectionBuilder);
      expect(builder.toString()).toMatchSnapshot();
      expect(mockSectionBuilder).not.toBeCalled();
    });
  });

  test('should work', () => {
    expect(
      new CodeBuilder({
        boil_body: 'new God().magic();',
      })
        .insertCode("import path from 'path';")
        .insertCode("import fs from 'fs'")
        .addManualSection('custom_imports', (builder) => builder)
        .insertCode(
          `class Steam extends Water {
              constructor() {
                this.boil();
              }

              boil() {`,
        )
        .addManualSection('boil_body', (builder) => builder.insertCode('this.temp = 100;'))
        .insertCode(`}`)
        .insertCode(`}`)
        .format()
        .toString(),
    ).toMatchSnapshot();
  });
});
