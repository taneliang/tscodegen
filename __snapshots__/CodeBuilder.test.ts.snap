// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeBuilder addManualSection should generate manual section using a new builder 1`] = `
"
/* BEGIN MANUAL SECTION mansec */
console.log(\\"New manual section\\");
/* END MANUAL SECTION */
"
`;

exports[`CodeBuilder addManualSection should retain existing manual section content if present 1`] = `
"
/* BEGIN MANUAL SECTION mansec */
console.log(\\"Existing manual section\\");
/* END MANUAL SECTION */
"
`;

exports[`CodeBuilder should work 1`] = `
"import path from \\"path\\";
import fs from \\"fs\\";
/* BEGIN MANUAL SECTION custom_imports */

/* END MANUAL SECTION */
class Steam extends Water {
  constructor() {
    this.boil();
  }

  boil() {
    /* BEGIN MANUAL SECTION boil_body */
    new God().magic();
    /* END MANUAL SECTION */
  }
}
"
`;
